network/containerizer/bridgepolicy.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package containerizer
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/set"
+
+	"github.com/juju/juju/network"
+	// Used for some constants and things like LinkLayerDevice[Args]
+	"github.com/juju/juju/state"
+)
+
+var logger = loggo.GetLogger("juju.network.containerizer")
+
+// BridgePolicy defines functionality that helps us create and define bridges
+// for guests inside of a host machine, along with the creation of network
+// devices on those bridges for the containers to use.
+// Ideally BridgePolicy would be defined outside of the 'state' package as it
+// doesn't deal directly with DB content, but not quite enough of State is exposed
+type BridgePolicy struct {
+}
+
+// Machine describes either a host machine, or a container machine. Either way
+// *state.Machine should fulfill the interface in real code.
+type Machine interface {
+	Id() string
+	AllSpaces() (set.Strings, error)
+	LinkLayerDevicesForSpaces([]string) (map[string][]*state.LinkLayerDevice, error)
+}
+
+var _ Machine = (*state.Machine)(nil)
+
+// Container describes the extra attributes we need to be able to update the
+// devices for a container.
+// *state.Machine should fulfill the interface.
+type Container interface {
+	Machine
+	DesiredSpaces() (set.Strings, error)
+	SetLinkLayerDevices(...state.LinkLayerDeviceArgs) error
+}
+
+var _ Container = (*state.Machine)(nil)
+
+// inferContainerSpaces tries to find a valid space for the container to be
+// on. This should only be used when the container itself doesn't have any
+// valid constraints on what spaces it should be in.
+// If this machine is in a single space, then that space is used. Else, if
+// the machine has the default space, then that space is used.
+// If neither of those conditions is true, then we return an error.
+func (p *BridgePolicy) inferContainerSpaces(m Machine, containerId, defaultSpaceName string) (set.Strings, error) {
+	hostSpaces, err := m.AllSpaces()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("container %q not qualified to a space, host machine %q is using spaces %v",
```

Comment:
Do we want to quote the space name output as we have done elsewhere? (IIRC, we had a function to do this.)